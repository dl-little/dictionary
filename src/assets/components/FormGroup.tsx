import { ReactElement } from 'react';

const FormGroup: React.FC<IFormGroup> = (props) => {
  const { input, handleInput } = props;

  const renderInput = (input: InputGroup) => {
    let inputEl: ReactElement<HTMLInputElement | HTMLSelectElement>;
    const defaultViteVar = 'VITE_' + input.default;
    const sharedProps = {
      className: 'border-gray-400 border',
      onChange: handleInput,
      required: true,
      ...(!!input.default &&
        !!import.meta.env[defaultViteVar] && {
          defaultValue: import.meta.env[defaultViteVar],
        }),
    };

    switch (input.type) {
      case 'number':
        inputEl = (
          <input
            type="text"
            inputMode="numeric"
            pattern="[0-9]*"
            id={input.id}
            {...sharedProps}
          />
        );
        break;
      case 'select':
        inputEl = (
          <select id={input.id} {...sharedProps}>
            <option hidden key="blank" value="-1">
              Choose which type of faker object to create
            </option>
            {input.options?.map((option) => {
              return (
                <option key={option.value} value={option.value}>
                  {option.display}
                </option>
              );
            })}
          </select>
        );
        break;
      default:
        inputEl = <input type={input.type} id={input.id} {...sharedProps} />;
    }

    return inputEl;
  };

  return (
    <div className="w-full flex flex-col">
      <label className="font-bold" htmlFor={input.id}>
        {input.label}
      </label>
      {renderInput(input)}
    </div>
  );
};

export default FormGroup;
