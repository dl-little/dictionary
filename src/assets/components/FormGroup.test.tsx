import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { describe, it, expect } from 'vitest';
import axe from 'axe-core';
import FormGroup from './FormGroup';

const setup = (group: JSX.Element) => {
  return {
    user: userEvent.setup(),
    ...render(group),
  };
};

describe('Functional - FormGroup', () => {
  it('renders the correct input of type text', async () => {
    render(
      <FormGroup
        input={{
          id: 'text_input',
          name: 'Text Input',
          label: 'Text Input',
          type: 'text',
        }}
      />
    );
    const input = await screen.findByRole('textbox', { name: /Text Input/i });
    expect(input).toMatchSnapshot();
  });

  it('renders the correct input of type number', async () => {
    render(
      <FormGroup
        input={{
          id: 'number_input',
          name: 'Number Input',
          label: 'Number Input',
          type: 'number',
        }}
      />
    );
    const input = (await screen.findByRole('textbox', {
      name: /Number Input/i,
    })) as HTMLInputElement;
    expect(input).toMatchSnapshot();
  });

  it('uses pattern validation for input of type number', async () => {
    const { user } = setup(
      <FormGroup
        input={{
          id: 'number_input',
          name: 'Number Input',
          label: 'Number Input',
          type: 'number',
        }}
      />
    );
    const input = (await screen.findByRole('textbox', {
      name: /Number Input/i,
    })) as HTMLInputElement;

    await user.type(input, 'Non-numerical entry.');
    expect(input.checkValidity()).toBeFalsy();

    await user.click(input);
    await user.clear(input);
    await user.type(input, '12');
    expect(input.checkValidity()).toBeTruthy();
  });
});

describe('Accessibility - FormGroup', () => {
  it('does not render with any accessibility violations', async () => {
    const { user } = setup(
      <FormGroup
        input={{
          id: 'number_input',
          name: 'Number Input',
          label: 'Number Input',
          type: 'number',
        }}
      />
    );

    const input = (await screen.findByRole('textbox', {
      name: /Number Input/i,
    })) as HTMLInputElement;

    const results = await axe.run(input);
    expect(results.violations.length).toBe(0);
  });
});
