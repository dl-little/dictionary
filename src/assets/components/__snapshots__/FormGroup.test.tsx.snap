// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Functional - FormGroup > renders the correct input of type number 1`] = `
<input
  class="border-gray-400 border"
  id="number_input"
  inputmode="numeric"
  name="Number Input"
  pattern="[0-9]*"
  required=""
  type="text"
/>
`;

exports[`Functional - FormGroup > renders the correct input of type text 1`] = `
<input
  class="border-gray-400 border"
  id="text_input"
  name="Text Input"
  required=""
  type="text"
/>
`;
